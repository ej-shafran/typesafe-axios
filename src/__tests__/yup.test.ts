import { taxios } from "..";

import * as yup from "yup";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = yup.object({
  id: yup.string().required(),
  username: yup.string().required(),
  email: yup.string().required(),
  address: yup.object({
    street: yup.string().required(),
    zipcode: yup.string(),
  }),
  followers: yup.array().of(
    yup.object({
      id: yup.string().required(),
      role: yup.string().oneOf(["admin", "guest"]),
    })
  ),
});

type TYPE = yup.InferType<typeof SCHEMA>;
function ensure(_: TYPE) { }

describe("yup interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);
    const { data } = await taxios(SCHEMA).get(UTILS.API);

    expect(data).toEqual(UTILS.VALID);
    ensure(data);
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    try {
      await taxios(SCHEMA).get(UTILS.API);
    } catch (error) {
      expect(JSON.stringify(error, null, 2)).toMatchSnapshot();
    }
  });
});
