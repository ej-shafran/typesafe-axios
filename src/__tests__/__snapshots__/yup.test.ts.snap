// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yup interop throws a validation error when the data is invalid 1`] = `
"{
  "value": {
    "id": "SOME_ID",
    "username": "SOME_USERNAME",
    "address": "ABC123 St.",
    "followers": []
  },
  "path": "address",
  "type": "typeError",
  "errors": [
    "address must be a \`object\` type, but the final value was: \`\\"ABC123 St.\\"\`."
  ],
  "params": {
    "value": "ABC123 St.",
    "originalValue": "ABC123 St.",
    "path": "address",
    "spec": {
      "strip": false,
      "strict": false,
      "abortEarly": true,
      "recursive": true,
      "nullable": false,
      "optional": true,
      "coerce": true
    },
    "type": "object"
  },
  "inner": [],
  "name": "ValidationError",
  "message": "address must be a \`object\` type, but the final value was: \`\\"ABC123 St.\\"\`."
}"
`;
